clippy_lints/src/copies.rs

```diff
@@ -352,6 +355,7 @@ fn scan_block_for_eq(cx: &LateContext<'_>, conds: &[&Expr<'_>], blocks: &[&Block
             // Any `i` from `blocks.windows(2)` will exist in `conds`, but `i+1` may not exist on the last iteration.
             if !matches!(conds[i].kind, ExprKind::Let(..));
             if !matches!(conds.get(i + 1).map(|e| &e.kind), Some(ExprKind::Let(..)));
+            if !matches!(cx.tcx.item_name(macro_call.def_id).as_str(), "todo"); // Took this from src/panic_unimplemented.rs
```

Comment:
I would like to add a condition here which is checking if the bodies of `if/else` all contain `todo!()`. If they do, the linter wouldn't trigger. I naively copied-pasted some code from `src/panic_unimplemented.rs` to start from code which is already handling `todo!()`. I would like to know how I can check for `todo!()` in the block which is currently scanned when this function is called. Could someone guide me with this?

Reply:
To check for a macro call you'll need to start from a `Span` (I would be the span of the last statement in this case). With the span you can use `span.ctxt().outer_expn_data()` which will get you an `ExpnData` which contains the `DefId` for the macro which created the statement. From there you can use `clippy_utils::match_def_path`. You'll probably need to add a path for `todo` in `clippy_utils::paths`.

`ExpnData` also has the span for the call site which you'll need if `todo` immediately calls another macro. You can use it to walk through the whole chain of macro calls.

Reply:
Thank you @Jarcho for your message. I've read about paths and by doing so, I also learned about diagnostic items. According to this [documentation](https://github.com/rust-lang/rust/blob/43d9f3859e0204e764161ee085a360274b5f3e9a/src/tools/clippy/doc/common_tools_writing_lints.md#checking-for-a-specific-type), it seems like relying on diagnostic items is preferred over paths. I will try to implement a solution with diagnostic items.